### ISSUE TITLE
[BUG] Setting MediaElement.Source to an embedded video does not work? No matter how it is set?

### ISSUE THREAD

**Issue by jonmdev** [CommunityToolkit/Maui#1433]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Did you read the "Reporting a bug" section on Contributing file?

- [X] I have read the "Reporting a bug" section on Contributing file: https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#reporting-a-bug

### Current Behavior

According to the [documentation](https://learn.microsoft.com/en-us/dotnet/communitytoolkit/maui/views/mediaelement), one should be able to play an embedded video as per the following:

- _To enable a media file for playback from the local resources add the file to the Resources/Raw folder of you .NET MAUI project. When a file is added in the root, the URI would be embed://MyFile.mp4._
- _In code, please use MediaSource.FromResource() and MediaSource.FromFile() respectively. Using these methods, you can omit the the embed:// and filesystem:// prefixes. The rest of the path should be the same._

However, this does not work at all for an embedded Resource video. I am able to successfully play videos from a URL as follows:

```
            MediaElement mediaElement = new();
            mediaElement.Source = "https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4";
            mediaElement.Play();
```
I am also able to play files in Windows build from a direct local file path as follows:

```
            MediaElement mediaElement = new();
            mediaElement.Source = MediaSource.FromFile("H:\\filePath\\BigBuckBunny.mp4");
            mediaElement.Play();
```

However, no matter what I try, I cannot get a `MediaSource.FromResource()` to work. I have placed the video in the general `Resources/Raw` folder as requested:

<img width="236" alt="mediaplayer resources 1" src="https://github.com/CommunityToolkit/Maui/assets/43948450/f8b4d9b1-6891-44db-9e81-80b31f901754">

I have also set it as an embedded resource:

<img width="236" alt="mediaplayer resources 2" src="https://github.com/CommunityToolkit/Maui/assets/43948450/2d308f66-169e-4a83-a581-1dec82142660">

And if I run the following code to output all embedded resources, I get as follows:

```
        private void debugResources() {
            foreach (string currentResource in System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceNames()) {
                Debug.WriteLine(currentResource);
            }
        }
```

Output: `MauiBuild.Resources.Raw.BigBuckBunny.mp4` showing it is being embedded in the project.

Yet trying to run the video with any of the following for source does not work:

`            mediaElement.Source = MediaSource.FromResource("MauiBuild.Resources.Raw.BigBuckBunny.mp4");`
`            mediaElement.Source = MediaSource.FromResource("BigBuckBunny.mp4");`
`            mediaElement.Source = new Uri("embed://MauiBuild.Resources.Raw.BigBuckBunny.mp4");`
`            mediaElement.Source = new Uri("embed://BigBuckBunny.mp4");`

Should not one of these methods work as suggested by documentation?

### Expected Behavior

At least one of the attempted methods to set the `mediaElement.Source` from the embedded resource should work.

### Steps To Reproduce

1. Create new Maui project in Visual Studio 2022 using .NET 7.0 by File > New

2. Add Maui Community Toolkit and MediaElement packages

3. Download video from `https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4`, put in common `Resources\Raw` folder with properties set to Embedded Resource.

4. Create anywhere in page and it will play:

```
            MediaElement mediaElement = new();
            mediaElement.Source = "https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4";
            mediaElement.Play();
```

5. Try to do the same but set Source to the Embedded video in `Resources\Raw` using any of the above methods and nothing will play.

### Link to public reproduction project repository

https://github.com/jonmdev/MediaElement-Resource-Bug

### Environment

```markdown
- .NET MAUI CommunityToolkit: 5.3.0 
- .NET Maui MediaElement 2.0.0
- OS: Windows 10, Android
- .NET MAUI: 7.0
```

**Comment by jonmdev** [comment]
### Steps to Reproduce

This takes at most 32 lines of code and 1 minute to reproduce. 

1. Start new project by File > New in Visual Studio 2022 with .NET 7.0, call it "MediaElement Resource Bug"
2. Install MediaElement nuget: https://www.nuget.org/packages/CommunityToolkit.Maui.MediaElement
3. Add line `.UseMauiCommunityToolkitMediaElement()` to the builder chain of MauiProgram.cs
4. Download this 1 MB test video to `Resources/Raw` and set properties to "Embedded Resource": https://sample-videos.com/video123/mp4/720/big_buck_bunny_720p_1mb.mp4
5. Copy and paste the following code to replace all of App.xaml.cs:

```
using System.Diagnostics;
using CommunityToolkit.Maui.Views;

namespace MediaElement_Resource_Bug {
    public partial class App : Application {
        public App() {
            InitializeComponent();

            MediaElement mediaElement = new ();
            mediaElement.Source = MediaSource.FromUri("https://sample-videos.com/video123/mp4/720/big_buck_bunny_720p_1mb.mp4"); //alternative video link: https://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4
            //mediaElement.Source = "https://sample-videos.com/video123/mp4/720/big_buck_bunny_720p_1mb.mp4";
            //mediaElement.Source = MediaSource.FromResource("big_buck_bunny_720p_1mb.mp4");
            //mediaElement.Source = MediaSource.FromResource("MediaElement_Resource_Bug.Resources.Raw.big_buck_bunny_720p_1mb.mp4");
            mediaElement.ShouldShowPlaybackControls = false;
            mediaElement.ShouldAutoPlay = true;
            mediaElement.Play();
            
            MainPage = new ContentPage();
            (MainPage as ContentPage).Content = mediaElement;

            //Debug Resources to test if embedded:
            TapGestureRecognizer gesture = new();
            mediaElement.GestureRecognizers.Add(gesture);
            gesture.Tapped += delegate {
                foreach (string currentResource in System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceNames()) {
                    Debug.WriteLine(currentResource);
                }
            };
        }
    }
}
```

That's it. You will see that the project will play the video fine when using a URL either directly or `FromUri` but will not play from an embedded resource. You can also click the screen to debug out the embedded resources to see it is being included.

**Comment by brminnick** [comment]
Thank you @jonmdev. However, we request a reproduction sample uploaded to a public source repository like GitHub or GitLab.

> We always request a reproduction sample, and that's not to make your life hard or anything like that... The reason of having a reproduction is to save us time to indentify and fix the bug. Create a new project; download the MCT NuGet; write UI; ViewModel; create a service; run it. As you can see it's a lot of timing consuming for us that we could spend fixing the actual issue. So PLEASE create a small reproduction project, upload it on GitHub or GitLab and paste the link in the issue.
>
> We don't accept .zip files as reproduction samples, for security reasons. So if you send us a .zip file we will kindly ask you to upload it to GitHub or GitLab and share the link
>
> And most important: Please, help us to help you ❤️


https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#reporting-a-bug

**Comment by bijington** [comment]
@jonmdev we typically ask for a reproduction project as it helps us to both allocate our time more efficiently, and also follow precisely what has been done.

I cannot say that I have tried running on Windows but I can confirm that I can see the same issue on Android using our sample app at: https://github.com/CommunityToolkit/Maui/blob/d2ee835201b4d7056e6b99e0684816101b847d9b/samples/CommunityToolkit.Maui.Sample/Pages/Views/MediaElement/MediaElementPage.xaml.cs#L174

I see the following in the output window which may help provide context:
> CommunityToolkit.Maui.Sample.Pages.Views.MediaElementPage: Information: Media State Changed. Old State: Buffering, New State: Failed
CommunityToolkit.Maui.Sample.Pages.Views.MediaElementPage: Information: Media failed. Error: Error codename: ERROR_CODE_IO_FILE_NOT_FOUND, Error message: Source error
MediaManager: Error: Error codename: ERROR_CODE_IO_FILE_NOT_FOUND, Error message: Source error

It does work on macOS.

**Comment by jonmdev** [comment]
Well I think at this point we have all agreed the function is unfortunately broken at least.

**Comment by jonmdev** [comment]
I tested in .NET 8.0 and tried testing in iOS. In iOS with .NET 8.0 it spits out the following error when trying to load from resource and crashes (can't start up app):

```
2023-10-06 04:31:11.219 Xamarin.PreBuilt.iOS[1888:906616] Could not resolve assembly System.Private.DataContractSerialization.resources, Version=7.0.0.0, Culture=en, PublicKeyToken=null. Details: Could not load file or assembly '/var/mobile/Containers/Data/Application/FFB54D43-E1D6-43CA-93FA-35F77FA82BDF/Documents/MediaElement Resource Bug.content/System.Private.DataContractSerialization.resources.dll' or one of its dependencies.

2023-10-06 04:31:11.217 Xamarin.PreBuilt.iOS[1888:906616] Could not resolve assembly System.Private.DataContractSerialization.resources, Version=7.0.0.0, Culture=en-CA, PublicKeyToken=null. Details: Could not load file or assembly '/var/mobile/Containers/Data/Application/FFB54D43-E1D6-43CA-93FA-35F77FA82BDF/Documents/MediaElement Resource Bug.content/System.Private.DataContractSerialization.resources.dll' or one of its dependencies.

[0:] An error occurred: 'Value cannot be null. (Parameter 'url')'. Callstack: '   at ObjCRuntime.ThrowHelper.ThrowArgumentNullException(String argumentName)
   at ObjCRuntime.NativeObjectExtensions.GetNonNullHandle(INativeObject self, String argumentName)
   at AVFoundation.AVAsset.FromUrl(NSUrl url)
   at CommunityToolkit.Maui.Core.Views.MediaManager.PlatformUpdateSource()
   at CommunityToolkit.Maui.Core.Views.MediaManager.UpdateSource()
   at CommunityToolkit.Maui.Core.Handlers.MediaElementHandler.MapSource(MediaElementHandler handler, MediaElement MediaElement)
   at Microsoft.Maui.PropertyMapper`2.<>c__DisplayClass5_0[[CommunityToolkit.Maui.Views.MediaElement, CommunityToolkit.Maui.MediaElement, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null],[CommunityToolkit.Maui.Core.Handlers.MediaElementHandler, CommunityToolkit.Maui.MediaElement, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].<Add>b__0(IElementHandler h, IElement v)
   at Microsoft.Maui.PropertyMapper.UpdatePropertyCore(String key, IElementHandler viewHandler, IElement virtualView)
   at Microsoft.Maui.PropertyMapper.UpdateProperties(IElementHandler viewHandler, IElement virtualView)
   at Microsoft.Maui.Handlers.ElementHandler.SetVirtualView(IElement view)
```

That is with trying to load via: `mediaElement.Source = MediaSource.FromResource("MediaElement_Resource_Bug.Resources.Raw.big_buck_bunny_720p_1mb.mp4");`
or `mediaElement.Source = MediaSource.FromResource("big_buck_bunny_720p_1mb.mp4");` - Both are the same and it claims this doesn't exist even though it does as proven by `System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceNames()`.

By contrast `mediaElement.Source = MediaSource.FromUri("https://sample-videos.com/video123/mp4/720/big_buck_bunny_720p_1mb.mp4"); ` loads the app okay but it is not playing on iOS either from this.

So to summarize for URL playback:

- Windows can play from URL
- Android can play from URL
- iOS can play from URL 

For embedded resources:

- Windows cannot play from embedded resource
- Android cannot play from embedded resource
- iOS crashes when it tries to play from embedded resource

In no cases (.NET 7 or .NET 8, and with Windows/iOS/Android) can `MediaSource.FromResource` find the resources that are listed by `System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceNames()`.

**Comment by bijington** [comment]
Thanks for the detail. I know they is an ongoing PR for .NET8 support but we don't officially support that yet. That being said it is clear we have an issue with embedded resource on the current version.

It is interesting that iOS fails whereas macOS does not. I am not able to look at this before next week but I can try to dig deeper then

**Comment by michaelonz** [comment]
I am also experiencing this issue

**Comment by bijington** [comment]
I haven't progressed very far with this issue only to note that if I make use of the `AssetManager` from Android I can see the file is listed as an asset:

```csharp
var assets = PlayerView?.Context?.Assets;
```

**Comment by ne0rrmatrix** [comment]
Workaround for now is:   ```mediaElement.Source = MediaSource.FromResource("Assets/BigBuckBunny.mp4");``` 
in csproj: 
````
<ItemGroup>
  <Resource Include="Resources\Raw\BigBuckBunny.mp4" />
</ItemGroup>
````
Use above for android and u can load assets from resources as video.

**Comment by bijington** [comment]
The detail above your comment shows the PR that has fixed the issue. We just need to get a release out with the fix in

**Comment by saramgsilva** [comment]
Hello, 

I am trying to use the MediaElement in my project, but nothing is happening

I have it
![image](https://github.com/CommunityToolkit/Maui/assets/2135829/731d6ea6-51ce-4f02-af2e-fd87907481c6)

and it

        <ItemGroup>
	     <Resource Include="Resources\Raw\bg_cuted.mp4" />
	</ItemGroup>


![image](https://github.com/CommunityToolkit/Maui/assets/2135829/f2ac291d-ca10-44ba-8bc5-b914776075e1)

![image](https://github.com/CommunityToolkit/Maui/assets/2135829/495ce446-578b-4c27-829a-141878da2c63)




the events are not fired.
Which permissions is necessary for it? What I am doing wrong in my project when I made a simple sample in the same way and the workround works.

**Comment by ne0rrmatrix** [comment]
@saramgsilva can you try this in your csproj file? 
```csharp
 <ItemGroup>
     <Resource Include="Resources\Raw\bg_cuted.mp4" />
     <CopyToOutputDirectory>Always</CopyToOutputDirectory>
</ItemGroup>
```

**Comment by saramgsilva** [comment]
Thanks is working link my sample works.

**Comment by jonmdev** [comment]
Thanks guys it does seem to be working now. I tested with .NET 8 and current version of Toolkit and I can get Windows and Android both to play the embedded video.

**Comment by hungchips** [comment]
This issue still exists as of 2024 when I posted this comment with the latest NET8 MAUI, and cannot be solved by the above methods. I have tried searching and testing many, many times on Android for many days.
+ Embeded Resource
+ MauiAsset
+ Method MediaSource.FromResource
+ Method MediaSource.FromFile
+ WebView and Javascript can see and read the `Embeded resouce` or `MauiAsset` files correctly, But MediaElement 4.0.1 always reports that `file NotFound`.

Visual Studio 17.10.3
MAUI Toolkit 9.0.2, MediaElement 4.0.1, MAUI Control 8.0.61

**Comment by bijington** [comment]
@hungchips this issue is closed. If you are seeing a similar issue can you please open a new issue?

**Comment by jonmdev** [comment]
> This issue still exists as of 2024 when I posted this comment with the latest NET8 MAUI, and cannot be solved by the above methods. I have tried searching and testing many, many times on Android for many days.
> 
>     * Embeded Resource
> 
>     * MauiAsset
> 
>     * Method MediaSource.FromResource
> 
>     * Method MediaSource.FromFile
> 
>     * WebView and Javascript can see and read the `Embeded resouce` or `MauiAsset` files correctly, But MediaElement 4.0.1 always reports that `file NotFound`.
> 
> 
> Visual Studio 17.10.3 MAUI Toolkit 9.0.2, MediaElement 4.0.1, MAUI Control 8.0.61

One idea - try debugging out the resources links like this and be sure you are copying and pasting it correctly into your code.

```
        private void debugResources() {
            foreach (string currentResource in System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceNames()) {
                Debug.WriteLine(currentResource);
            }
        }
```
It should work. I am able to play videos from resources this way.

I am curious: How do you access your resources from inside a webview html/javascript? What is the path? Is it the relative file folder or what?

For example, debugging the above might get:

`ProjectName.Resources.Raw.BigBuckBunny.mp4`

which should work for MediaElement. How would you access this from javascript/webview inside the project?

**Comment by hungchips** [comment]
Visual Studio 17.10.3 MAUI Toolkit 9.0.2, MediaElement 4.0.1, MAUI Control 8.0.61

I am really sorry for my ignorance in applying the methods for older versions.

NET8 MAUI we can put the .mp3 or .mp4 file in any folder.
For example: ProjectName/HTML/Music/BlackIsTheCourlor.mp3.
Set BuildAction to MauiAsset
Then do MediaElement.Source = MediaSource.FromResource("HTML/Music/BlackIsTheCourlor.mp3")

Thanks everyone. It really worked.
